Iss hipcms 648

Musaevt opened the pull request.

Musaevt committed:
Added Tags Controller

Musaevt committed:
Added Api calls
Post api/Tag
Get api/Tag

Musaevt committed:
Added Api calls
--Update By id =>  PUT /api/tag/{id}
--Get By Id => Get/api/tag/{id}

Musaevt committed:
Merge branch 'develop' into iss-hipcms-648

Musaevt committed:
Added call Delete api/{id}

Musaevt committed:
Merge branch 'HIPCMS-632' into iss-hipcms-648

Musaevt committed:
Added Image field to TagUpdated
Added updating tag in db

SvenEV commented on HiP-DataStore.Model/Rest/TagArgs.cs:
```diff
@@ -0,0 +1,20 @@
+    public class TagArgs
+    {
+        [Required]
+        public string Tille { get; set; }
```
Typo (should be "Title")

SvenEV commented on HiP-DataStore/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,197 @@
+        [ProducesResponseType(typeof(int), 200)]
+        [ProducesResponseType(400)]
+        [ProducesResponseType(422)]
+        [ProducesResponseType(409)]
```
Method should return 409 if a tag with the given name already exists. This does not seem to be implemented yet.

SvenEV commented on HiP-DataStore/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,197 @@
+
+            var query = _db.Database.GetCollection<Tag>(ResourceType.Tag.Name).AsQueryable();
+
+            var tag = query.Where(x => x.Id == id).FirstOrDefault();
```
instead of ".Where(condition).FirstOrDefault()" you can use ".FirstOrDefault(condition)"

SvenEV commented on HiP-DataStore.Model/Rest/TagUpdateArgs.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace PaderbornUniversity.SILab.Hip.DataStore.Model.Rest
+{
+    public class TagUpdateArgs
```
I discussed updates and PUT requests with Patrick (see latest comments on page "REST API for mobile content" on Confluence). We came to the conclusion that PUT requests shouldn't be used to update individual fields of an entity, but rather to replace an entity completely. Thus, PUT requests should not only specify values for fields that should be changed, but also for fields that do not change. In other words, POST and PUT always have exactly the same kinds of parameters, so TagUpdateArgs is redundant and we can just use TagArgs instead.

SvenEV commented on HiP-DataStore/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,197 @@
+        [ProducesResponseType(400)]
+        [ProducesResponseType(404)]
+        [ProducesResponseType(422)]
+        public async Task<IActionResult> UpdateById(int id,[FromBody]TagUpdateArgs args)
```
UpdateById must not only add a reference to the image (if specified), but also remove any existing references (for example, if a PUT request changes the image ID from 2 to 5, remove the old reference to 2, then update the tag, then add a new reference to 5).

SvenEV commented on HiP-DataStore/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,197 @@
+            if (!_entityIndex.Exists(ResourceType.Tag, id))
+                return NotFound();
+
+            if (!_referencesIndex.IsUsed(ResourceType.Tag, id))
```
Inverted if condition (remove the "!" from `!_referencesIndex.IsUsed(...)`)

SvenEV commented on HiP-DataStore/Core/ReadModel/CacheDatabaseManager.cs:
```diff
@@ -119,13 +121,53 @@ private void OnEventAppeared(EventStoreCatchUpSubscription subscription, Resolve
+                        bsonDoc = new BsonDocument("$set", bsonDoc);
+                        _db.GetCollection<Tag>(ResourceType.Tag.Name).UpdateOne(x => x.Id == e.Id, bsonDoc);
+                    break;
 
```
Case for TagDeleted is missing

SvenEV commented on HiP-DataStore/Core/WriteModel/MediaIndex.cs:
```diff
@@ -11,35 +11,84 @@ namespace PaderbornUniversity.SILab.Hip.DataStore.Core.WriteModel
+            }
+        }
+       
+        public bool IsImage(int id)
```
* Missing locks in IsImage and IsAudio
* `? true : false` is redundant

SvenEV commented on HiP-DataStore.Model/Entity/Tag.cs:
```diff
@@ -9,6 +9,8 @@ public class Tag : ContentBase
 
         public string Title { get; set; }
 
+        public bool IsUsed { get; set; }
```
Property should be removed. Use `tag.Referencees.Count > 0` to check if a tag is in use.

Musaevt committed:
Changed to "Title"
instead of ".Where(condition).FirstOrDefault()"  using ".FirstOrDefault(condition)"
Added case for TagDeleted
Added Locks to IsImage and IsAudio
Property IsUsed is removed

Musaevt committed:
Implemented deliting old referencies in Update method
Changed Arguments for Update method
Added StatusCode 209 to POST and PUT

Musaevt committed:
deleted redundant directives

Musaevt committed:
deleted redundant property name

patrickrobrecht committed:
Merge branch 'develop' into iss-hipcms-648

SvenEV commented on HiP-DataStore/Core/WriteModel/TagIndex.cs:
```diff
@@ -0,0 +1,44 @@
+{
+    public class TagIndex : IDomainIndex
+    {
+        private readonly Dictionary<int, TagInfo> _tag = new Dictionary<int, TagInfo>();
```
What do we need the event IDs for? I think a HashSet<string> should suffice, no need for a dictionary. Event IDs are unique anyway, so entries in the dictionary will never be updated or deleted.

Musaevt replied:
It isn`t an event ID. It is a Tag ID which we have to create/update/delete

SvenEV replied:
My fault. "ev.Id" is of course a tag ID, not an event ID, so everything is fine. Sorry!

SvenEV commented on HiP-DataStore/Core/WriteModel/TagIndex.cs:
```diff
@@ -0,0 +1,44 @@
+                    break;
+                case TagDeleted ev:
+                    lock (_lockObject)
+                        _tag.Remove(ev.Id);
```
This won't work, see comment above. The TagCreated event and the TagDeleted event will have different event IDs, so this line never removes anything from the dictionary.

SvenEV committed:
Merge branch 'develop' into iss-hipcms-648

SvenEV committed:
Fixed compiler errors after merging 'develop'
